### PR TITLE
Ability to show exported variables only when viewed within their scene

### LINKED ISSUE #10024: Annotation to only display property if node is root of current scene
### Describe the project you are working on

A top-down 2D shooter, as well as a rhythm-centric turn-based RPG (although these specifics aren't relevant here)

### Describe the problem or limitation you are having in your project

Before making a proposal, this feature idea was discussed here: https://github.com/godotengine/godot-proposals/discussions/10017

I like to use `@export`ed properties on my scripts to link sub-nodes to them. For example, I might have a Player scene like this:

CleanShot 2024-06-23 at 16 54 54@2x

With the following script, I can drag and drop the nodes into the property slots, and then access them in the Player script:
```gdscript
extends Node2D

@export var sprite: Sprite2D
@export var sound_player: AudioStreamPlayer2D
@export var health: int = 10
...
```
CleanShot 2024-06-23 at 16 55 34@2x

If I place this scene in another scene, those slots still take up space, despite the fact that they should be "fixed" at this level (if I want to change them, I'll go into the Player scene to do that).

CleanShot 2024-06-23 at 16 59 41@2x

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I am proposing that functionality be added to GDScript that allows the user to specify that **a property should only be drawn in the inspector if it is on the root node of the currently-edited scene**. This will make it easier to adhere to the "black box" philosophy of scenes in Godot, and hide extraneous cluttering information from the user as they work with scenes on a higher level.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

For now, I propose adding an annotation to GDScript, `@export_if_root`, which is added _in addition to_ an existing `@export` annotation (or one of its many variants). When this annotation is added to a property, the property is only considered to be drawn if it is the root node of the scene currently being edited.

GDScript code of the current implementation:
```gdscript
extends Node2D

@export @export_if_root var sprite: Sprite2D
@export @export_if_root var sound_player: AudioStreamPlayer2D
@export @export_if_root var health: int = 10

func _process(_delta):
    print("Sprite = %s, sound player = %s, health = %s" % [sprite, sound_player, health])
```
This is currently implemented in my branch of Godot here: https://github.com/Meorge/godot/tree/export_if_root

**I recognize that this is _not_ necessarily the best exact syntax for this feature, and should be discussed before anything is merged.** Here are a few alternatives I thought of.
- `@export @if_root var my_property` - This reads a bit more naturally and takes up less space. However it may be confusing to see on its own, and/or incorrectly imply that it modifies the behavior of other decorators as well.
- `@export(true) var my_property` - Making this annotation into an argument for `@export` would certainly be the most succinct approach, but it would have to be added onto every variant of `@export`. On top of that, the lack of keyword arguments means that it's not immediately clear what the argument refers to. One could reasonably - yet incorrectly - assume that `@export(true)` means "export this variable" and `@export(false)` means "don't export this variable". 

### If this enhancement will not be used often, can it be worked around with a few lines of script?

As users take further advantage of nested scenes, this enhancement will be used more and more.

The behavior can be accomplished with pure GDScript, but it feels rather clunky to me when combined with regular game code:
```gdscript
@tool
extends Node2D

@export var sprite: Sprite2D
@export var sound_player: AudioStreamPlayer2D
@export var health: int = 10

func _process(_delta):
    if Engine.is_editor_hint():
        return
    print("Sprite = %s, sound player = %s, health = %s" % [sprite, sound_player, health])

func _validate_property(property):
    var scene_root = EditorInterface.get_edited_scene_root()
    if property.name in ["sprite", "sound_player"] and scene_root.is_ancestor_of(self) and not scene_root.is_editable_instance(self):
        property.usage = PROPERTY_USAGE_NO_EDITOR
```

This has a few drawbacks, in my opinion:
1. When we add the `@tool` annotation to the script, making it run in the editor, we have to add checks in (at least) the `_ready()` and `_process()` functions to ensure that the actual script behavior only runs when the game is playing. This boilerplate feels unnecessary and bloated to me.
2. The logic inside of `_validate_property()` function isn't immediately intuitive or readable in my opinion. Additionally, requiring the property names to be written in an array means that if for some reason the user changes the property names, they'll have to go back and change them here as well. With an annotation on the variable itself, this wouldn't happen.

### Is there a reason why this should be core and not an add-on in the asset library?

I don't believe there is a way to create custom annotations or modify existing ones with how Godot and GDScript currently works. Thus, I think it would have to be core rather than an add-on.

### PR DESCRIPTION
**Pull request by Meorge**
This PR adds a new annotation to GDScript, `@export_if_root`, which when placed after an `@export` annotation, causes the annotated property to only show in the inspector if the node is the root of the current scene. This is useful for hiding irrelevant information when editing nested scenes.

There is a discussion thread at https://github.com/godotengine/godot-proposals/discussions/10017 and a formal proposal at https://github.com/godotengine/godot-proposals/issues/10024. More details about the reasoning behind the feature are provided there (especially in the proposal). Before merging anything, I think it may be wise for us to refine the syntax.

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/10024*

### CODE REVIEW

**Comment by theraot** [comment]
This is technically not a review, but I can tell you right now that you should add documentation for `PROPERTY_USAGE_SHOW_IF_ROOT`. But perhaps you want to wait until you iron out the other details. I'll also make a post in the proposal shortly.

**Comment by AThousandShips** [comment]
There's some formatting errors, please run [formatting](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html) and [add to documentation](https://docs.godotengine.org/en/latest/contributing/documentation/updating_the_class_reference.html) before we can run CI on this

**Comment by Meorge** [comment]
I can certainly do that! For the documentation specifically, should I go ahead and write up the documentation for the implementation as it is currently, even if the details (such as the name of the annotation) are likely to change before being merged in?

**Comment by AThousandShips** [comment]
You should at least generate them, but you can leave a temporary description until things are decided on

**Comment by Meorge** [comment]
Alrighty, thanks! Will try to get those done today then 😄

File: core/core_constants.cpp [hunk 1]
```diff
@@ -710,7 +710,7 @@ void register_global_constants() {
 	BIND_CORE_BITFIELD_FLAG(PROPERTY_USAGE_EDITOR_BASIC_SETTING);
 	BIND_CORE_BITFIELD_FLAG(PROPERTY_USAGE_READ_ONLY);
 	BIND_CORE_BITFIELD_FLAG(PROPERTY_USAGE_SECRET);
-
+	BIND_CORE_BITFIELD_FLAG(PROPERTY_USAGE_SHOW_IF_ROOT);
 	BIND_CORE_BITFIELD_FLAG(PROPERTY_USAGE_DEFAULT);
```
Comment (AThousandShips) [inline core/core_constants.cpp:714]:
```suggestion

	BIND_CORE_BITFIELD_FLAG(PROPERTY_USAGE_DEFAULT);
```
Keep an empty line separating these for readability

File: core/object/object.h [hunk 2]
```diff
@@ -120,7 +120,7 @@ enum PropertyUsageFlags {
 	PROPERTY_USAGE_EDITOR_BASIC_SETTING = 1 << 27, //for project or editor settings, show when basic settings are selected.
 	PROPERTY_USAGE_READ_ONLY = 1 << 28, // Mark a property as read-only in the inspector.
 	PROPERTY_USAGE_SECRET = 1 << 29, // Export preset credentials that should be stored separately from the rest of the export config.
-
+	PROPERTY_USAGE_SHOW_IF_ROOT = 1 << 30, // Only show a property in the inspector if the node is the current root.
 	PROPERTY_USAGE_DEFAULT = PROPERTY_USAGE_STORAGE | PROPERTY_USAGE_EDITOR,
```
Comment (AThousandShips) [inline core/object/object.h:124]:
```suggestion

	PROPERTY_USAGE_DEFAULT = PROPERTY_USAGE_STORAGE | PROPERTY_USAGE_EDITOR,
```

File: modules/gdscript/doc_classes/@GDScript.xml [hunk 3]
```diff
@@ -577,6 +577,17 @@
 				[/codeblock]
 			</description>
 		</annotation>
+		<annotation name="@export_if_root">
+			<return type="void" />
+			<description>
+				When placed between an [code]@export[/code] annotation and the exported property, causes the exported property to only be displayed in the Inspector dock if the currently-selected node is the root node of the current scene.
```
Comment (AThousandShips) [inline modules/gdscript/doc_classes/@GDScript.xml:583]:
```suggestion
				When placed between an [annotation @export] annotation and the exported property, causes the exported property to only be displayed in the Inspector dock if the currently-selected node is the root node of the current scene.
```

**Comment by AThousandShips** [comment]
I feel this annotation is far too specific, I can think of few cases when I'd want something to show up *only* when the node is the root

Honestly I'd personally be more interested in when it *isn't* the root but instead is in an instanced scene, that makes far more sense to me

But I feel this is far too specific and also far too easy to work around in custom code to warrant a custom annotation, you can simply do this with `_validate_property` and `EditorInterface.get_edited_scene_root()` (though that might not work easily in exported projects, but that's something to solve with other features like editor only compiled code, not something against that workaround)

**Comment by AThousandShips** [comment]
> I'm not sure at which point it's acceptable to take a proposal and try to implement it

There really isn't any limit for that, the only thing to consider is that you might be wasting time if you try to implement a proposal before it's gotten good support and ends up rejected because it's not in high enough demand, or having to redesign it multiple times because the discussion changes and other details are focused on

So beyond that letting a proposal sit until some stronger consensus has been reached it's up to you, but jumping on a proposal before it's been discussed deeply enough can risk making a PR when the discussion would otherwise have led to dismissing the idea or some workaround being found or some different solution being decided on

**Comment by Meorge** [comment]
Closing this as there's not a lot of interest in it, and/or if the `private` access modifier ends up going forward, a similar function could potentially be a part of that.

**Pull request closed without merging**